// @payloadcms/plugin-form-builder/dist/collections/Forms/fields.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var DynamicFieldSelector_1 = require("./DynamicFieldSelector");
var DynamicPriceSelector_1 = require("./DynamicPriceSelector");
var name = {
    name: 'name',
    label: 'Name (lowercase, no special characters)',
    type: 'text',
    required: true,
};
var label = {
    name: 'label',
    label: 'Label',
    type: 'text',
    localized: true,
};
var required = {
    name: 'required',
    label: 'Required',
    type: 'checkbox',
};
var width = {
    name: 'width',
    label: 'Field Width (percentage)',
    type: 'number',
};
var Select = {
    slug: 'select',
    labels: {
        singular: 'Select',
        plural: 'Select Fields',
    },
    fields: [
        {
            type: 'row',
            fields: [
                __assign(__assign({}, name), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, label), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        {
            type: 'row',
            fields: [
                __assign(__assign({}, width), { admin: {
                        width: '50%',
                    } }),
                {
                    name: 'defaultValue',
                    label: 'Default Value',
                    localized: true,
                    type: 'text',
                    admin: {
                        width: '50%',
                    },
                },
            ],
        },
        {
            name: 'options',
            label: 'Select Attribute Options',
            type: 'array',
            labels: {
                singular: 'Option',
                plural: 'Options',
            },
            fields: [
                {
                    type: 'row',
                    fields: [
                        {
                            name: 'label',
                            label: 'Label',
                            type: 'text',
                            required: true,
                            localized: true,
                            admin: {
                                width: '50%',
                            },
                        },
                        {
                            name: 'value',
                            label: 'Value',
                            type: 'text',
                            required: true,
                            admin: {
                                width: '50%',
                            },
                        },
                    ],
                },
            ],
        },
        required,
    ],
};
var Text = {
    slug: 'text',
    labels: {
        singular: 'Text',
        plural: 'Text Fields',
    },
    fields: [
        {
            type: 'row',
            fields: [
                __assign(__assign({}, name), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, label), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        {
            type: 'row',
            fields: [
                __assign(__assign({}, width), { admin: {
                        width: '50%',
                    } }),
                {
                    name: 'defaultValue',
                    label: 'Default Value',
                    type: 'text',
                    localized: true,
                    admin: {
                        width: '50%',
                    },
                },
            ],
        },
        required,
    ],
};
var TextArea = {
    slug: 'textarea',
    labels: {
        singular: 'Text Area',
        plural: 'Text Area Fields',
    },
    fields: [
        {
            type: 'row',
            fields: [
                __assign(__assign({}, name), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, label), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        {
            type: 'row',
            fields: [
                __assign(__assign({}, width), { admin: {
                        width: '50%',
                    } }),
                {
                    name: 'defaultValue',
                    label: 'Default Value',
                    localized: true,
                    type: 'text',
                    admin: {
                        width: '50%',
                    },
                },
            ],
        },
        required,
    ],
};
var Number = {
    slug: 'number',
    labels: {
        singular: 'Number',
        plural: 'Number Fields',
    },
    fields: [
        {
            type: 'row',
            fields: [
                __assign(__assign({}, name), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, label), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        {
            type: 'row',
            fields: [
                __assign(__assign({}, width), { admin: {
                        width: '50%',
                    } }),
                {
                    name: 'defaultValue',
                    label: 'Default Value',
                    type: 'number',
                    admin: {
                        width: '50%',
                    },
                },
            ],
        },
        required,
    ],
};
var Email = {
    slug: 'email',
    labels: {
        singular: 'Email',
        plural: 'Email Fields',
    },
    fields: [
        {
            type: 'row',
            fields: [
                __assign(__assign({}, name), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, label), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        width,
        required,
    ],
};
var State = {
    slug: 'state',
    labels: {
        singular: 'State',
        plural: 'State Fields',
    },
    fields: [
        {
            type: 'row',
            fields: [
                __assign(__assign({}, name), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, label), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        width,
        required,
    ],
};
var Country = {
    slug: 'country',
    labels: {
        singular: 'Country',
        plural: 'Country Fields',
    },
    fields: [
        {
            type: 'row',
            fields: [
                __assign(__assign({}, name), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, label), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        width,
        required,
    ],
};
var Checkbox = {
    slug: 'checkbox',
    labels: {
        singular: 'Checkbox',
        plural: 'Checkbox Fields',
    },
    fields: [
        {
            type: 'row',
            fields: [
                __assign(__assign({}, name), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, label), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        {
            type: 'row',
            fields: [
                __assign(__assign({}, width), { admin: {
                        width: '50%',
                    } }),
                __assign(__assign({}, required), { admin: {
                        width: '50%',
                    } }),
            ],
        },
        {
            name: 'defaultValue',
            label: 'Default Value',
            type: 'checkbox',
        },
    ],
};
var Payment = function (fieldConfig) {
    var paymentProcessorField = null;
    if (fieldConfig === null || fieldConfig === void 0 ? void 0 : fieldConfig.paymentProcessor) {
        paymentProcessorField = __assign({ type: 'select', options: [], name: 'paymentProcessor', label: 'Payment Processor' }, fieldConfig.paymentProcessor);
    }
    var fields = {
        slug: 'payment',
        labels: {
            singular: 'Payment',
            plural: 'Payment Fields',
        },
        fields: [
            {
                type: 'row',
                fields: [
                    __assign(__assign({}, name), { admin: {
                            width: '50%',
                        } }),
                    __assign(__assign({}, label), { admin: {
                            width: '50%',
                        } }),
                ],
            },
            {
                type: 'row',
                fields: [
                    __assign(__assign({}, width), { admin: {
                            width: '50%',
                        } }),
                    {
                        name: 'basePrice',
                        type: 'number',
                        label: 'Base Price',
                        admin: {
                            width: '50%',
                        },
                    },
                ],
            },
            paymentProcessorField,
            {
                name: 'priceConditions',
                labels: {
                    singular: 'Price Condition',
                    plural: 'Price Conditions',
                },
                type: 'array',
                label: 'Price Conditions',
                fields: [
                    {
                        name: 'fieldToUse',
                        type: 'text',
                        admin: {
                            components: {
                                Field: DynamicFieldSelector_1.DynamicFieldSelector,
                            },
                        },
                    },
                    {
                        name: 'condition',
                        label: 'Condition',
                        type: 'select',
                        defaultValue: 'hasValue',
                        options: [
                            {
                                value: 'hasValue',
                                label: 'Has Any Value'
                            },
                            {
                                value: 'equals',
                                label: 'Equals'
                            },
                            {
                                value: 'notEquals',
                                label: 'Does Not Equal'
                            }
                        ]
                    },
                    {
                        name: 'valueForCondition',
                        label: 'Value',
                        type: 'text',
                        admin: {
                            condition: function (_, _a) {
                                var condition = _a.condition;
                                return condition === 'equals' || condition === 'notEquals';
                            }
                        }
                    },
                    {
                        name: 'operator',
                        type: 'select',
                        defaultValue: 'add',
                        options: [
                            {
                                value: 'add',
                                label: 'Add'
                            },
                            {
                                value: 'subtract',
                                label: 'Subtract'
                            },
                            {
                                value: 'multiply',
                                label: 'Multiply'
                            },
                            {
                                value: 'divide',
                                label: 'Divide'
                            }
                        ]
                    },
                    {
                        name: 'valueType',
                        label: 'Value Type',
                        type: 'radio',
                        admin: {
                            width: '100%',
                        },
                        defaultValue: 'static',
                        options: [
                            {
                                label: 'Static Value',
                                value: 'static'
                            },
                            {
                                label: 'Value Of Field',
                                value: 'valueOfField'
                            }
                        ]
                    },
                    {
                        name: 'valueForOperator',
                        label: 'Value',
                        type: 'text',
                        admin: {
                            components: {
                                Field: DynamicPriceSelector_1.DynamicPriceSelector,
                            },
                        },
                    },
                ]
            },
            required,
        ].filter(Boolean)
    };
    return fields;
};
var Message = {
    slug: 'message',
    labels: {
        singular: 'Message',
        plural: 'Message Blocks',
    },
    fields: [
        {
            name: 'message',
            type: 'richText',
            localized: true,
        },
    ],
};
exports.default = {
    select: Select,
    checkbox: Checkbox,
    text: Text,
    textarea: TextArea,
    email: Email,
    message: Message,
    number: Number,
    country: Country,
    state: State,
    payment: Payment
};
