// @payloadcms/plugin-nested-docs/dist/fields/breadcrumbs.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
var createBreadcrumbsField = function (relationTo, overrides) {
    if (overrides === void 0) { overrides = {}; }
    return (__assign({ name: 'breadcrumbs', type: 'array', fields: __spreadArray([
            {
                name: 'doc',
                type: 'relationship',
                relationTo: relationTo,
                maxDepth: 0,
                admin: {
                    disabled: true,
                },
            },
            {
                type: 'row',
                fields: [
                    {
                        name: 'url',
                        label: 'URL',
                        type: 'text',
                        admin: {
                            width: '50%',
                        },
                    },
                    {
                        name: 'label',
                        type: 'text',
                        admin: {
                            width: '50%',
                        },
                    },
                ],
            }
        ], (overrides === null || overrides === void 0 ? void 0 : overrides.fields) || [], true), admin: __assign({ readOnly: true }, (overrides === null || overrides === void 0 ? void 0 : overrides.admin) || {}) }, overrides || {}));
};
exports.default = createBreadcrumbsField;
