// @payloadcms/plugin-cloud-storage/dist/adapters/azure/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.azureBlobStorageAdapter = void 0;
var storage_blob_1 = require("@azure/storage-blob");
var staticHandler_1 = require("./staticHandler");
var generateURL_1 = require("./generateURL");
var handleDelete_1 = require("./handleDelete");
var handleUpload_1 = require("./handleUpload");
var webpack_1 = require("./webpack");
var azureBlobStorageAdapter = function (_a) {
    var connectionString = _a.connectionString, containerName = _a.containerName, baseURL = _a.baseURL, allowContainerCreate = _a.allowContainerCreate;
    return function (_a) {
        var collection = _a.collection, prefix = _a.prefix;
        var blobServiceClient = storage_blob_1.BlobServiceClient.fromConnectionString(connectionString);
        var containerClient = blobServiceClient.getContainerClient(containerName);
        return {
            handleUpload: (0, handleUpload_1.getHandleUpload)({
                collection: collection,
                containerClient: containerClient,
                allowContainerCreate: allowContainerCreate,
                prefix: prefix,
            }),
            handleDelete: (0, handleDelete_1.getHandleDelete)({ collection: collection, containerClient: containerClient }),
            generateURL: (0, generateURL_1.getGenerateURL)({ containerName: containerName, baseURL: baseURL }),
            staticHandler: (0, staticHandler_1.getHandler)({ containerClient: containerClient, collection: collection }),
            webpack: webpack_1.extendWebpackConfig,
        };
    };
};
exports.azureBlobStorageAdapter = azureBlobStorageAdapter;
